### PR TITLE
feat(client): Expose metrics to the client

### LINKED ISSUE #12875: Expose Metrics in Prisma Client
Once the query engine has implemented metrics we need to expose it via the Prisma Client Api

Depends on https://github.com/prisma/prisma/issues/12874
Epic:https://github.com/prisma/client-planning/issues/2

### PR DESCRIPTION
**Pull request by SevInf**
Close #12875

### CODE REVIEW

File: packages/client/tests/functional/metrics/tests.ts
Comment (SevInf) [inline packages/client/tests/functional/metrics/tests.ts:41]: So, here, rather than doing snaphots and introducing dependency on engine implementation details, we assert that returned metrics has the data and general properties of the data kind of make sense. This should make test more resilient to engine internal changes, which will cause the changes to the metrics.

File: packages/client/tests/functional/metrics/tests.ts
Comment (SevInf) [inline packages/client/tests/functional/metrics/tests.ts:77]: This nested describe is here only for `afterAll`'s sake - we need to ensure that second client gets disconnected even if the test fails.

**Review by garrensmith — COMMENTED** [review]
This is looking good. Easy code to follow.

File: packages/client/tests/functional/metrics/tests.ts
Comment (SevInf) [reply packages/client/tests/functional/metrics/tests.ts:98]: Yes, `toEqual` does deep comparison.

**Comment by SevInf** [comment]
Now it uses updated engines and is ready for review

File: packages/engine-core/src/data-proxy/DataProxyEngine.ts
Comment (millsp) [reply packages/engine-core/src/data-proxy/DataProxyEngine.ts]: Should we perhaps error at generation time like we do for `interactiveTransactions` with `--data-proxy`?

File: packages/client/src/runtime/core/metrics/index.ts
Comment (millsp) [inline packages/client/src/runtime/core/metrics/index.ts:1]: Is this file/pattern in case we want to have better/centralized exports?

File: packages/client/tests/functional/metrics-disabled/tests.ts
Comment (millsp) [inline packages/client/tests/functional/metrics-disabled/tests.ts:9]: Do you mean that we need to have dev-time types/intellisense?

File: packages/client/tests/functional/metrics/_matrix.ts
Comment (millsp) [inline packages/client/tests/functional/metrics/_matrix.ts:11]: I really like that we could test the engineTypes this way! However, the CI already does that by running all the tests with binary and again with library, and it does it via the `PRISMA_CLIENT_ENGINE_TYPE` env var and via `getClientEngineType` in the code. tl;dr; You don't need to add anything and can remove both engine types.

**Review by millsp — COMMENTED** [review]
Very nice work 👏

File: packages/client/tests/functional/metrics-disabled/tests.ts
Comment (SevInf) [reply packages/client/tests/functional/metrics-disabled/tests.ts:9]: Yea, basically, right now it always highlighted as an error in editors, since `prisma` has `any` type there and any can have anything. It works when you actually run the tests, since there `prisma` is properly typed by then. Having proper types during development will fix this. I'd maybe do it separately as a part of testing project

File: packages/client/src/runtime/core/metrics/index.ts
Comment (SevInf) [reply packages/client/src/runtime/core/metrics/index.ts:1]: So, here I want two things:

1. Centralized place for importing anything metrics-related, hence separate folder with `index.ts`
2. Still being to import to open `MetricsClient.ts` file by name, hence `index.ts` doing just a re-export

Admittedly, "anything metrics-related" right now is only `MetricsClient` class and it will quite likely stay that way, so I guess we could remove this file. What would you prefer?

File: packages/engine-core/src/data-proxy/DataProxyEngine.ts
Comment (SevInf) [reply packages/engine-core/src/data-proxy/DataProxyEngine.ts]: Ok, i'll wait until #13561  is merged then

**Comment by garrensmith** [comment]
I've noticed a bug if I use the client in binary mode. If I do:
```
const prisma = new PrismaClient({
  log: ["info", "error", "query", "warn"],
});
let run = async () => {
  let metrics = await prisma.$metrics.prometheus();
};
```

I get the error:
```
/Users/garren/dev/prisma/reports/ttt/node_modules/@prisma/client/runtime/index.js:43171
    throw new Error("Connection has not been opened");
          ^
Error: Connection has not been opened
    at assertHasPool (/Users/garren/dev/prisma/reports/ttt/node_modules/@prisma/client/runtime/index.js:43171:11)
    at Connection.raw (/Users/garren/dev/prisma/reports/ttt/node_modules/@prisma/client/runtime/index.js:43197:5)
    at Connection.post (/Users/garren/dev/prisma/reports/ttt/node_modules/@prisma/client/runtime/index.js:43215:17)
    at BinaryEngine.metrics (/Users/garren/dev/prisma/reports/ttt/node_modules/@prisma/client/runtime/index.js:43977:44)
    at MetricsClient.prometheus (/Users/garren/dev/prisma/reports/ttt/node_modules/@prisma/client/runtime/index.js:39905:25)
 ```

but if I do
```
const prisma = new PrismaClient({
  log: ["info", "error", "query", "warn"],
});
let run = async () => {
  await prisma.$connect();
  let metrics = await prisma.$metrics.prometheus();
};
```
Then it works. I'm guessing we need to make sure we open our HTTP client.

**Comment by SevInf** [comment]
@garrensmith fixed

File: packages/engine-core/src/data-proxy/DataProxyEngine.ts
Comment (janpio) [reply packages/engine-core/src/data-proxy/DataProxyEngine.ts]: You are right, this should indeed be handled as early as possible.

**Review by millsp — APPROVED** [review]

**Pull request merged**
